File: sdks/dart/lib/src/mcp.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'sdk_info.dart';

/// Streamable HTTP MCP client (sak338-c).
class SakMcpClient {
  SakMcpClient(String? baseUrl, {http.Client? httpClient})
      : baseUrl = _strip((baseUrl == null || baseUrl.trim().isEmpty)
            ? 'http://127.0.0.1:8080/mcp'
            : baseUrl),
        _http = httpClient ?? http.Client();

  final String baseUrl;
  final http.Client _http;

  String? token;
  bool autoInitialize = true;

  int _rpcId = 0;
  String? _sessionId;
  bool _initialized = false;

  String? get sessionId => _sessionId;

  static String _strip(String u) {
    var out = u;
    while (out.endsWith('/')) {
      out = out.substring(0, out.length - 1);
    }
    return out;
  }

  Future<dynamic> initialize() async {
    final result = await _rpc('initialize', {
      'protocolVersion': '2024-11-05',
      'capabilities': <String, dynamic>{},
      'clientInfo': {'name': 'swissarmynoife-dart', 'version': SdkInfo.version},
    });
    await _post({'jsonrpc': '2.0', 'method': 'notifications/initialized'},
        notification: true);
    _initialized = true;
    return result;
  }

  Future<String> ping() async {
    return _extractPingText(await _toolsCall('ping'));
  }

  Future<dynamic> toolsList() async {
    await _ensureSession();
    return _rpc('tools/list');
  }

  Future<dynamic> catalogList() async => _toolsCall('catalog_list');

  Future<void> _ensureSession() async {
    if (!autoInitialize || _initialized) return;
    await initialize();
  }

  Future<dynamic> _toolsCall(String name,
      [Map<String, dynamic> arguments = const {}]) async {
    await _ensureSession();
    return _rpc('tools/call', {'name': name, 'arguments': arguments});
  }

  Future<dynamic> _rpc(String method,
      [Map<String, dynamic> params = const {}]) async {
    _rpcId++;
    final res = await _post({
      'jsonrpc': '2.0',
      'id': _rpcId,
      'method': method,
      'params': params,
    });
    final body = jsonDecode(res.body);
    _captureSession(res, body);
    if (body is Map && body.containsKey('error')) {
      final err = body['error'];
      final msg = err is Map ? (err['message'] ?? err) : err;
      throw StateError('MCP $method failed: $msg');
    }
    if (body is Map && body.containsKey('result')) return body['result'];
    return body;
  }

  Future<http.Response> _post(Map<String, dynamic> payload,
      {bool notification = false}) async {
    final headers = <String, String>{
      'content-type': 'application/json',
      'accept': 'application/json, text/event-stream',
    };
    if (token != null && token!.isNotEmpty) {
      headers['authorization'] = 'Bearer $token';
    }
    if (_sessionId != null && _sessionId!.isNotEmpty) {
      headers['mcp-session-id'] = _sessionId!;
    }
    final res = await _http.post(
      Uri.parse(baseUrl),
      headers: headers,
      body: jsonEncode(payload),
    );
    if (notification && (res.statusCode == 200 || res.statusCode == 202)) {
      return res;
    }
    if (res.statusCode < 200 || res.statusCode >= 300) {
      throw StateError('${res.statusCode}: ${res.body}');
    }
    return res;
  }

  void _captureSession(http.Response res, dynamic body) {
    if (_sessionId == null || _sessionId!.isEmpty) {
      final sid = res.headers['mcp-session-id'];
      if (sid != null && sid.trim().isNotEmpty) {
        _sessionId = sid.trim();
      }
    }
    if ((_sessionId == null || _sessionId!.isEmpty) && body is Map) {
      final fromBody = _sessionIdFromBody(body);
      if (fromBody != null) _sessionId = fromBody;
    }
  }

  static String? _sessionIdFromBody(Map body) {
    for (final key in ['sessionId', 'session_id', 'mcp-session-id']) {
      final v = body[key];
      if (v is String && v.trim().isNotEmpty) return v.trim();
    }
    final result = body['result'];
    if (result is Map) {
      for (final key in ['sessionId', 'session_id', 'mcp-session-id']) {
        final v = result[key];
        if (v is String && v.trim().isNotEmpty) return v.trim();
      }
    }
    return null;
  }

  static String _extractPingText(dynamic result) {
    if (result is String) return result;
    if (result is Map) {
      final content = result['content'];
      if (content is List) {
        for (final item in content) {
          if (item is Map && item['text'] is String) {
            return item['text'] as String;
          }
        }
      }
    }
    return '$result';
  }
}

File: sdks/go/mcp.go
package swissarmynoife

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

const (
	defaultMCPURL         = "http://127.0.0.1:8080/mcp"
	mcpProtocolVersion    = "2024-11-05"
	mcpSessionHeader      = "mcp-session-id"
	mcpAccept             = "application/json, text/event-stream"
)

// McpClient is a Streamable HTTP MCP client (sak330-c).
type McpClient struct {
	BaseURL         string
	Token           string
	HTTPClient      *http.Client
	AutoInitialize  bool
	rpcID           int
	sessionID       string
	initialized     bool
}

// NewMcpClient returns an MCP client. baseURL may be empty for the default.
// AutoInitialize defaults to true.
func NewMcpClient(baseURL string) *McpClient {
	u := strings.TrimRight(baseURL, "/")
	if u == "" {
		u = defaultMCPURL
	}
	return &McpClient{
		BaseURL:        u,
		HTTPClient:     &http.Client{Timeout: 30 * time.Second},
		AutoInitialize: true,
	}
}

// SessionID returns the negotiated mcp-session-id, if any.
func (m *McpClient) SessionID() string { return m.sessionID }

func (m *McpClient) authHeaders() http.Header {
	h := make(http.Header)
	h.Set("Content-Type", "application/json")
	h.Set("Accept", mcpAccept)
	if m.Token != "" {
		h.Set("Authorization", "Bearer "+m.Token)
	}
	if m.sessionID != "" {
		h.Set(mcpSessionHeader, m.sessionID)
	}
	return h
}

func (m *McpClient) post(payload map[string]any, notification bool) (*http.Response, error) {
	raw, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodPost, m.BaseURL, bytes.NewReader(raw))
	if err != nil {
		return nil, err
	}
	req.Header = m.authHeaders()
	res, err := m.HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	if notification && (res.StatusCode == 200 || res.StatusCode == 202) {
		return res, nil
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		body, _ := io.ReadAll(res.Body)
		res.Body.Close()
		return nil, fmt.Errorf("%d: %s", res.StatusCode, string(body))
	}
	return res, nil
}

func (m *McpClient) captureSession(res *http.Response, body any) {
	if m.sessionID == "" {
		if sid := strings.TrimSpace(res.Header.Get(mcpSessionHeader)); sid != "" {
			m.sessionID = sid
		}
	}
	if m.sessionID == "" {
		if sid := sessionIDFromBody(body); sid != "" {
			m.sessionID = sid
		}
	}
}

func sessionIDFromBody(body any) string {
	obj, ok := body.(map[string]any)
	if !ok {
		return ""
	}
	for _, key := range []string{"sessionId", "session_id", "mcp-session-id"} {
		if v, ok := obj[key].(string); ok && strings.TrimSpace(v) != "" {
			return strings.TrimSpace(v)
		}
	}
	if result, ok := obj["result"].(map[string]any); ok {
		for _, key := range []string{"sessionId", "session_id", "mcp-session-id"} {
			if v, ok := result[key].(string); ok && strings.TrimSpace(v) != "" {
				return strings.TrimSpace(v)
			}
		}
	}
	return ""
}

func (m *McpClient) rpc(method string, params map[string]any) (any, error) {
	if params == nil {
		params = map[string]any{}
	}
	m.rpcID++
	payload := map[string]any{
		"jsonrpc": "2.0",
		"id":      m.rpcID,
		"method":  method,
		"params":  params,
	}
	res, err := m.post(payload, false)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	raw, err := io.ReadAll(res.Body)
	if err != nil {
		return nil, err
	}
	var body any
	if err := json.Unmarshal(raw, &body); err != nil {
		return nil, err
	}
	m.captureSession(res, body)
	obj, ok := body.(map[string]any)
	if !ok {
		return body, nil
	}
	if errObj, has := obj["error"]; has {
		msg := fmt.Sprint(errObj)
		if em, ok := errObj.(map[string]any); ok {
			if m0, ok := em["message"].(string); ok {
				msg = m0
			}
		}
		return nil, fmt.Errorf("MCP %s failed: %s", method, msg)
	}
	if result, has := obj["result"]; has {
		return result, nil
	}
	return body, nil
}

// Initialize negotiates a Streamable HTTP session.
func (m *McpClient) Initialize() (any, error) {
	result, err := m.rpc("initialize", map[string]any{
		"protocolVersion": mcpProtocolVersion,
		"capabilities":    map[string]any{},
		"clientInfo": map[string]any{
			"name":    "swissarmynoife-go",
			"version": "0.1.0",
		},
	})
	if err != nil {
		return nil, err
	}
	res, err := m.post(map[string]any{
		"jsonrpc": "2.0",
		"method":  "notifications/initialized",
	}, true)
	if err != nil {
		return nil, err
	}
	res.Body.Close()
	m.initialized = true
	return result, nil
}

func (m *McpClient) ensureSession() error {
	if !m.AutoInitialize || m.initialized {
		return nil
	}
	_, err := m.Initialize()
	return err
}

func (m *McpClient) toolsCall(name string, arguments map[string]any) (any, error) {
	if err := m.ensureSession(); err != nil {
		return nil, err
	}
	if arguments == nil {
		arguments = map[string]any{}
	}
	return m.rpc("tools/call", map[string]any{
		"name":      name,
		"arguments": arguments,
	})
}

// Ping calls tools/call ping and returns extracted text.
func (m *McpClient) Ping() (string, error) {
	result, err := m.toolsCall("ping", nil)
	if err != nil {
		return "", err
	}
	return extractPingText(result), nil
}

// ToolsList calls MCP tools/list.
func (m *McpClient) ToolsList() (any, error) {
	if err := m.ensureSession(); err != nil {
		return nil, err
	}
	return m.rpc("tools/list", nil)
}

// CatalogList calls tools/call catalog_list.
func (m *McpClient) CatalogList() (any, error) {
	return m.toolsCall("catalog_list", nil)
}

func extractPingText(result any) string {
	if s, ok := result.(string); ok {
		return s
	}
	obj, ok := result.(map[string]any)
	if !ok {
		return fmt.Sprint(result)
	}
	content, ok := obj["content"].([]any)
	if !ok {
		return fmt.Sprint(result)
	}
	for _, item := range content {
		im, ok := item.(map[string]any)
		if !ok {
			continue
		}
		if text, ok := im["text"].(string); ok {
			return text
		}
	}
	return fmt.Sprint(result)
}
